### PR TITLE
Enables `azd build` command and with hooks

### PR DESCRIPTION
**Pull request by wbreza**
- Adds `azd build` to `azd help` docs
- Adds project & service hooks for `build` command and lifecycle hooks

### CODE REVIEW

File: cli/azd/cmd/testdata/TestUsage-azd-build.snap [hunk 1]
```diff
@@ -0,0 +1,20 @@
+
+Builds the application's code. (Beta)
+
+Usage
+  azd build <service> [flags]
+
+Flags
+        --all                	: Deploys all services that are listed in azure.yaml
```
Comment (jongio) [inline cli/azd/cmd/testdata/TestUsage-azd-build.snap:8]:
```suggestion
        --all                	: Builds all services that are listed in azure.yaml
```

Comment (jongio) [inline cli/azd/cmd/testdata/TestUsage-azd-build.snap:8]: It defaults to all true right?

File: cli/azd/cmd/testdata/TestUsage-azd.snap [hunk 2]
```diff
@@ -7,6 +7,7 @@ Usage
 Commands
   Configure and develop your app
     auth     	: Authenticate with Azure.
+    build    	: Builds the application's code. (Beta)
```
Comment (jongio) [inline cli/azd/cmd/testdata/TestUsage-azd.snap:10]: Why (Beta)?

File: cli/azd/cmd/testdata/TestUsage-azd.snap [hunk 3]
```diff
@@ -7,6 +7,7 @@ Usage
...
     config   	: Manage azd configurations (ex: default Azure subscription, location).
     hooks    	: Develop, test and run hooks for an application. (Beta)
```
Comment (jongio) [inline cli/azd/cmd/testdata/TestUsage-azd.snap:12]: Not related, but why (Beta) on hooks?

**Comment by vhvb1989** [comment]
@jongio , IIRC, one of the reasons we decided not to support `azd build` was because it becomes complicated to automatically re-use command's output artifacts.
For example, if you run `azd build` and then `azd deploy`, will you be expecting to use the output from your previous `build` as part of deploy?  or are you expecting to do `build` again?

For `azd package` for example, we had to add `azd package --output FOO` and `azd deploy --from-package FOO` as a way to tell deploy what artifact to use.

So, adding `build` comes with a few questions about it which we must add to the help/docs for users to know what they can and can not do with `build` command.

I am not against having `azd build` public, but I am concern about the confusion about defining what's the purpose of the command.  Why and when do I want to use that command?

**Comment by jongio** [comment]
> @jongio , IIRC, one of the reasons we decided not to support `azd build` was because it becomes complicated to automatically re-use command's output artifacts. For example, if you run `azd build` and then `azd deploy`, will you be expecting to use the output from your previous `build` as part of deploy? or are you expecting to do `build` again?
> 
> For `azd package` for example, we had to add `azd package --output FOO` and `azd deploy --from-package FOO` as a way to tell deploy what artifact to use.
> 
> So, adding `build` comes with a few questions about it which we must add to the help/docs for users to know what they can and can not do with `build` command.
> 
> I am not against having `azd build` public, but I am concern about the confusion about defining what's the purpose of the command. Why and when do I want to use that command?

Scenario is the dev wants to build the entire solution to make sure everything builds, without having to wait for the packaging to complete.

We can expose it as-is now with some help docs and take feedback and modify it based on that feedback.

**Comment by ellismg** [comment]
> Scenario is the dev wants to build the entire solution to make sure everything builds, without having to wait for the packaging to complete.

Why does `azd` have to be the solution to this problem?  Is it because there is no good story for multi-language build orchestration here?  I assume if you were writing all of one language, you'd just use the build technology of that ecosystem?

File: cli/azd/cmd/testdata/TestUsage-azd.snap (hunk 2 above)
Comment (wbreza) [reply cli/azd/cmd/testdata/TestUsage-azd.snap:10]: It's new(ish) and may change.  We have an alpha, beta, GA strategy but we aren't applying it very consistently at the moment.  Open for any feedback here.

File: cli/azd/cmd/testdata/TestUsage-azd.snap (hunk 3 above)
Comment (wbreza) [reply cli/azd/cmd/testdata/TestUsage-azd.snap:12]: It was never promoted out of beta.  Happy to remove it when we're happy.

**Comment by wbreza** [comment]
> > Scenario is the dev wants to build the entire solution to make sure everything builds, without having to wait for the packaging to complete.
> 
> Why does `azd` have to be the solution to this problem? Is it because there is no good story for multi-language build orchestration here? I assume if you were writing all of one language, you'd just use the build technology of that ecosystem?

Why shouldn't `azd` help for the whole software development lifecycle?  As we move on to provide more local dev experiences a local build everything or build service foo is going to become very important in the day of the life of an azd developer.

`azd` should not be considered just a deployment tool.

**Comment by vhvb1989** [comment]
> Scenario is the dev wants to build the entire solution to make sure everything builds, without having to wait for the packaging to complete.
> 
> We can expose it as-is now with some help docs and take feedback and modify it based on that feedback.

@jongio , azd has an opinionated way of building projects. Even with `azd build`, you can't be sure everything builds as expected. What will you do for `azd build` for python services? 
For js, azd is just running `npm run build`, regardless of what the actual nodejs is doing. It might be using a different script to build, or it might just be a backend server that runs with nodejs without having a build script.
For Docker, we have been adding fields to azure.yaml to support more and more docker build settings, like --secrets, --build-args, --env-vars.  This makes sense for the specific build command you want to use for pushing your image to a registry to deploy to Azure. But you might want to build your container different when calling `azd build` for local development.
And same for DotNet.  Azd uses `Release` arg by default.

 What you would expect in a common DevOps world for -> `build the entire solution to make sure everything builds`  is to have a unique CI/CD pipeline for each project where you build and test the project as one alone.

**Pull request closed without merging**